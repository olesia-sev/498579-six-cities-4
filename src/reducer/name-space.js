const NameSpace = {
  DATA: `DATA`,
  APP: `APP`,
  USER: `USER`,
};

export {NameSpace};
