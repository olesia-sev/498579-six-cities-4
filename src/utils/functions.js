export const getRandomNumberInRange = (min, max) => {
  return Math.floor(Math.random() * (max - min + 1)) + min;
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};
