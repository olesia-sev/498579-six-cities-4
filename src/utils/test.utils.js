export const optionsAmount = 100;

export const testString = `Test string`;

export const emptyFunction = () => {};

export const cardDataArray = [
  {
    id: 10,
    img: `https://placedog.net/260/200?id=2`,
    images: [
      `https://placedog.net/260/200?id=3`,
      `https://placedog.net/260/200?id=33`,
      `https://placedog.net/260/200?id=12`,
      `https://placedog.net/260/200?id=10`,
      `https://placedog.net/260/200?id=11`,
      `https://placedog.net/260/200?id=15`,
    ],
    price: 150,
    title: `Title`,
    placeType: `Flat`,
    numberOfBedrooms: 2,
    maxGuests: 3,
    rating: 4,
    saved: true,
    premium: true,
    featuresInside: [
      `Wi-Fi`, `Heating`,
    ],
    hostName: `Pete`,
    hostAvatar: `https://placedog.net/260/200?id=5`,
    userPro: true,
    description: `Description`,
  },
  {
    id: 100,
    img: `https://placedog.net/260/200?id=3`,
    images: [
      `https://placedog.net/260/200?id=33`,
      `https://placedog.net/260/200?id=99`,
      `https://placedog.net/260/200?id=89`,
      `https://placedog.net/260/200?id=29`,
      `https://placedog.net/260/200?id=13`,
      `https://placedog.net/260/200?id=16`,
    ],
    price: 150,
    title: `Title test`,
    placeType: `Flat`,
    numberOfBedrooms: 1,
    maxGuests: 2,
    rating: 4,
    saved: false,
    premium: false,
    featuresInside: [
      `Wi-Fi`, `Heating`,
    ],
    hostName: `Pete`,
    hostAvatar: `https://placedog.net/260/200?id=57`,
    userPro: false,
    description: `Description`,
  }
];
