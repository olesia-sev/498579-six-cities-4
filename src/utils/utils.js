export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const SORT_TYPES = {
  POPULAR: `Popular`,
  TO_HIGH_PRICE: `Price: low to high`,
  TO_LOW_PRICE: `Price: high to low`,
  TOP_RATED: `Top rated first`,
};

export const getSortedOffers = (offers, activeSort) => {
  switch (activeSort) {
    case SORT_TYPES.POPULAR:
      return offers.slice();
    case SORT_TYPES.TO_HIGH_PRICE:
      return offers.slice().sort((a, b) => a.price - b.price);
    case SORT_TYPES.TO_LOW_PRICE:
      return offers.slice().sort((a, b) => b.price - a.price);
    case SORT_TYPES.TOP_RATED:
      return offers.slice().sort((a, b) => b.rating - a.rating);
    default:
      return offers.slice();
  }
};

export const MAX_REVIEWS_LENGTH = 10;

export const AppRoute = {
  LOGIN: `/login`,
  ROOT: `/`,
  FAVOURITES: `/favorites`,
  OFFER_DETAIL: `/offers/:id`,
};

export const RATING_STAR_WIDTH = 20;

export const MAX_IMAGES_AMOUNT = 6;
export const SLICE_BEGIN = 0;
