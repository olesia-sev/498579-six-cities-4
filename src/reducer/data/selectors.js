import {createSelector} from "reselect";
import {NameSpace} from "../name-space.js";

export const getOffers = (state) => {
  return state[NameSpace.DATA].offers;
};

export const getCurrentOffer = (state, offerId) => {
  return getOffers(state).find((offer) => offer.id === +offerId);
};

export const getActiveCityId = (state) => {
  return state[NameSpace.DATA].activeCityId;
};

export const getCities = (state) => {
  return state[NameSpace.DATA].cities;
};

export const getActiveCity = createSelector(
    getCities,
    getActiveCityId,
    (cities, activeCityId) => {
      return cities.find((city) => city.id === activeCityId);
    }
);

export const getFilteredOffers = createSelector(
    getOffers,
    getActiveCityId,
    (offers, activeCityId) => {
      return offers.filter((offer) => offer.cityId === activeCityId);
    }
);

export const getReviews = (state) => {
  return state[NameSpace.DATA].reviews;
};

export const getFavourites = (state) => {
  return state[NameSpace.DATA].favourites;
};

export const getFavoriteOffersIds = createSelector(
    getOffers,
    (offers) => {
      return offers.filter((offer) => offer.saved).map((offer) => offer.id);
    }
);

export const getFavoritesOffers = createSelector(
    [getFavourites, getFavoriteOffersIds],
    (favorites, offersIds) => {
      return favorites.filter((favoriteOffer) => offersIds.includes(favoriteOffer.id));
    }
);

export const getFavouritesLocations = createSelector(
    getFavoritesOffers,
    (result) => {
      return Array.from(new Set(result.map((it) => it.cityId)));
    }
);

export const offersNearbySelector = (state) => {
  return state[NameSpace.DATA].offersNearby;
};
