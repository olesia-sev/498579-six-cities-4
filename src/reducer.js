import {extend} from './utils/functions';
import cities from "./mocks/cities";
import offers from "./mocks/offers";

export const initialState = {
  activeCityId: 10,
  cities,
  offers,
};

const ActionType = {
  SET_ACTIVE_CITY: `SET_ACTIVE_CITY`,
};

const ActionCreator = {
  setActiveCity: (id) => ({
    type: ActionType.SET_ACTIVE_CITY,
    payload: id,
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_ACTIVE_CITY:
      return extend(state, {
        activeCityId: action.payload,
      });

    default: {
      return state;
    }
  }
};

export {reducer, ActionType, ActionCreator};
