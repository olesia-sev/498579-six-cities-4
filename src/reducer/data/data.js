import {AppRoute, extend, MAX_REVIEWS_LENGTH} from "../../utils/utils";
import {createOffers} from "../../adapters/offers";
import {createCity} from "../../adapters/cities";
import {mapReviews} from "../../adapters/reviews";
import {history} from "../../history";

export const initialState = {
  activeCityId: undefined,
  offers: [],
  cities: [],
  reviews: [],
};

const ActionType = {
  SET_ACTIVE_CITY: `SET_ACTIVE_CITY`,
  LOAD_OFFERS: `LOAD_OFFERS`,
  SET_CITIES: `SET_CITIES`,
  SET_REVIEWS: `SET_REVIEWS`,
  UPDATE_FAVOURITE: `UPDATE_FAVOURITE`,
};

const ActionCreator = {
  setActiveCity: (id) => ({
    type: ActionType.SET_ACTIVE_CITY,
    payload: id,
  }),
  loadOffers: (offers) => ({
    type: ActionType.LOAD_OFFERS,
    payload: offers,
  }),
  setCities: (cities) => ({
    type: ActionType.SET_CITIES,
    payload: cities,
  }),
  setReviews: (reviews) => ({
    type: ActionType.SET_REVIEWS,
    payload: reviews,
  }),
  updateFavourite: (offer) => ({
    type: ActionType.UPDATE_FAVOURITE,
    payload: offer,
  })
};

const Operation = {
  loadOffers: () => (dispatch, getState, api) => {
    return api.get(`/hotels`)
      .then((response) => {
        const loadedOffers = response.data.map(createOffers);
        const cities = response.data.reduce((acc, offer) => {
          const city = createCity(offer);
          acc[city.id] = city;
          return acc;
        }, {});
        dispatch(ActionCreator.setActiveCity(loadedOffers[0].cityId));
        dispatch(ActionCreator.loadOffers(loadedOffers));
        dispatch(ActionCreator.setCities(Object.values(cities)));
      })
      .catch((error) => {
        // eslint-disable-next-line no-console
        console.error(`[HOTELS ERROR]`, error.message);
      });
  },
  loadReviews: (offerId) => (dispatch, getState, api) => {
    return api.get(`/comments/${offerId}`)
      .then((response) => {
        dispatch(
            ActionCreator.setReviews(
                response.data
                  .slice(Math.max(response.data.length - MAX_REVIEWS_LENGTH, 0))
                  .map(mapReviews)
            )
        );
      })
      .catch((error) => {
        // eslint-disable-next-line no-console
        console.error(`[COMMENTS ERROR]`, error.message);
      });
  },
  postReview: (offerId, review) => (dispatch, getState, api) => {
    return api.post(`/comments/${offerId}`, review);
  },
  postFavourite: (offerId, status) => (dispatch, getState, api) => {
    return api.post(`/favorite/${offerId}/${status}`)
      .then((response) => {
        dispatch(ActionCreator.updateFavourite(response.data));
      })
      .catch((error) => {
        if (error.response.status === 401) {
          history.push(AppRoute.LOGIN);
        }
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_ACTIVE_CITY:
      return extend(state, {
        activeCityId: action.payload,
      });
    case ActionType.LOAD_OFFERS:
      return extend(state, {
        offers: action.payload
      });
    case ActionType.SET_CITIES:
      return extend(state, {
        cities: action.payload,
      });
    case ActionType.SET_REVIEWS:
      return extend(state, {
        reviews: action.payload,
      });
    case ActionType.UPDATE_FAVOURITE:
      return extend(state, {
        offers: state.offers.map((offer) => {
          if (offer.id === action.payload.id) {
            return createOffers(action.payload);
          }
          return offer;
        })
      });
  }
  return state;
};

export {Operation, reducer, ActionType, ActionCreator};
