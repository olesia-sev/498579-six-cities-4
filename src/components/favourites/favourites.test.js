import React from "react";
import {act, create} from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {cardDataArray} from "../../utils/test.utils";
import {NameSpace} from "../../reducer/name-space";
import Favourites from "./favourites";
import {BrowserRouter as Router} from "react-router-dom";
import {Operation} from "../../reducer/data/data";
import thunk from "redux-thunk";

const mockStore = configureStore([thunk]);

it(`Favourites should be rendered`, async () => {

  const store = mockStore({
    [NameSpace.DATA]: {
      activeCityId: `Amsterdam`,
      offers: cardDataArray,
      favourites: cardDataArray,
    },
    [NameSpace.USER]: {
      authorizationStatus: `AUTH`,
      authInfo: {fake: true},
    }
  });

  const spy = jest.spyOn(Operation, `loadFavourites`);
  spy.mockReturnValue(() => {
    return Promise.resolve();
  });

  let tree = create(null);
  await act(async () => {
    tree = create(
        <Router>
          <Provider store={store}>
            <Favourites />
          </Provider>
        </Router>
    );
  });

  expect(tree.toJSON()).toMatchSnapshot();
});
