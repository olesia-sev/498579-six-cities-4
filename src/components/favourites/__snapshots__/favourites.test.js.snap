// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favourites should be rendered 1`] = `
<div
  className="page
      
      "
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="/img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <span
                className="header__nav-link header__nav-link--profile"
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                  style={
                    Object {
                      "backgroundImage": "url('undefined')",
                    }
                  }
                />
                <a
                  className="header__user-name user__name"
                  href="/favorites"
                  onClick={[Function]}
                />
              </span>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--favorites
         
        "
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="/"
                  onClick={[Function]}
                >
                  <span>
                    Paris
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <article
                className="favorites__card place-card"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <div
                  className="place-card__mark"
                >
                  <span>
                    Premium
                  </span>
                </div>
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <a
                    href="/offers/10"
                    onClick={[Function]}
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height={110}
                      src="https://placedog.net/260/200?id=2"
                      width={150}
                    />
                  </a>
                </div>
                <div
                  className="favorites__card-info place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        150
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button button place-card__bookmark-button--active"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      href="/offers/10"
                      onClick={[Function]}
                    >
                      Title
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    Flat
                  </p>
                </div>
              </article>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
